test/test_cuda.py

```diff
@@ -3496,6 +3496,50 @@ class TestNamedTupleInput_1(NamedTuple):
             self.assertEqual(expected_a, x.a)
             self.assertEqual(expected_b, x.b)
 
+    @unittest.skipIf(not TEST_MULTIGPU, "Test needs multiple GPUs")
+    def test_gather_namedtuple(self):
+        # tests ability to gather a list of namedtuples and return a namedtuple where each
+        # element is of the expected tensor type.
+        fields = ['a', 'b']
+        TestNamedTupleInput_0 = collections.namedtuple('NamedTuple', fields)
+
+        num_gpus = torch.cuda.device_count()
+        a = torch.rand(num_gpus * 2, device=0)
+        b = torch.rand(num_gpus * 2, device=1)
+        out1 = TestNamedTupleInput_0(a, b)
+
+        a = torch.rand(num_gpus * 2, device=1)
+        b = torch.rand(num_gpus * 2, device=0)
+        out2 = TestNamedTupleInput_0(a, b)
+
+        outputs = [out1, out2]
+        out = scatter_gather.gather(outputs, 'cpu')  # test on cpu
```

Comment:
Should we have a test for GPU here, and a test for CPU for the one below so all paths are exercised?

Reply:
I'm not sure what you mean. Do you mean have either ```a``` or ```b``` tensor in the CPU and test if the gather function can gather from those devices?

Reply:
I meant that we should always have a test as follows:
`out = scatter_gather.gather(outputs, 0)`

basically similar to the test on line 3536, but for the `TestnamedtupleInput_0` type. And vice versa, the below test should also have `out = scatter_gather.gather(outputs, 'cpu')`?

Reply:
Ah I see. Those tests would ensure that a previous call to ```scatter_gather.gather``` does not affect the output of future calls to the function. I have added those tests!